test_apps/advection/Advection.cpp

```diff
@@ -0,0 +1,705 @@
+#include <iostream>
+#include "Advection.h"
+#include <fstream>
+
+using namespace flow;
+using namespace external;
+
+// Constructor;
+Advection::Advection() : _lowerAngle( 3.0f ), _upperAngle( 15.0f )
+{
+    _lowerAngleCos = glm::cos( glm::radians( _lowerAngle ) );
+    _upperAngleCos = glm::cos( glm::radians( _upperAngle ) );
+
+    for( int i = 0; i < 3; i++ )
+    {
+        _isPeriodic[i] = false;
+    }
+}
+
+void
+Advection::UseSeedParticles( const std::vector<Particle>& seeds )
+{
+    _streams.clear();
+    _streams.resize( seeds.size() );
+    for( size_t i = 0; i < seeds.size(); i++ )
+        _streams[i].push_back( seeds[i] );
+
+    _separatorCount.resize( seeds.size() );
+    for( auto& e : _separatorCount )
+        e = 0;
+}
+
+int
+Advection::CheckReady() const
+{
+
+    for( const auto& s : _streams )
+    {
+        if( s.size() < 1 )
+            return NO_SEED_PARTICLE_YET;
+    }
+
+    return 0;
+}
+
+int
+Advection::AdvectSteps( Field* velocity, float deltaT, size_t maxSteps, ADVECTION_METHOD method )
+{
+    int ready = CheckReady();
+    if( ready != 0 )
+        return ready;
+    // The particle advection process can be parallelized per particle
+    // Each stream represents a trajectory for a single particle
+    #pragma omp parallel
+    #pragma omp for
+    for(size_t streamIdx = 0; streamIdx < _streams.size(); streamIdx++)  // Process one stream at a time
+    {
+        auto& s = _streams.at(streamIdx);
+        size_t numberOfSteps = 0;
+        while(numberOfSteps < maxSteps)
+        {
+            // Check if the particle is inside of the volume.
+            // Also wrap it along periodic dimensions if enabled.
+            if( !velocity->InsideVolumeVelocity( s.back().time, s.back().location ) )
+            {
+                auto& p0 = s.back();
+                // Attempt to apply periodicity
+                bool locChanged = false;
+                auto loc = p0.location;
+                for( int i = 0; i < 3; i++ )    // correct coordinates in each periodic dimension
+                {
+                    if( _isPeriodic[i] )
+                    {
+                        loc[i] = _applyPeriodic( loc[i], _periodicBounds[i][0], _periodicBounds[i][1] );
+                        locChanged = true;
+                    }
+                }
+
+                if( !locChanged )   // no dimension is periodic
+                    break;       // skip this particle, since it's out of the volume
```

Comment:
I'm not sure why this line shows up again seemingly the same file... But still, is this break breaking an intended loop ?